### ISSUE TITLE
"git add -N" prevents restoring stashed changes.

### ISSUE THREAD

**Issue by henniss** [pre-commit/pre-commit#881]
To reproduce, start with this simple `pre-commit-config.yaml` in an otherwise empty repo:

```yaml
repos:
- repo: https://github.com/pre-commit/pre-commit-hooks
  rev: master
  hooks: 
    - id: end-of-file-fixer
```

The hook used doesn't really matter. end-of-file-fixer is just an example. 

Run the following:
```bash
echo "new" > newfile
echo "\n\n\n" > needs-fixing
git add -N newfile
# newfile is now staged as an empty file
git add needs-fixing
git commit -m "fix"
```

The following output is generated: 

```
[WARNING] Unstaged files detected.
[INFO] Stashing unstaged files to /home/henniss/.cache/pre-commit/patch1544663784.
Fix End of Files.........................................................Failed
hookid: end-of-file-fixer

Files were modified by this hook. Additional output:

Fixing needs-fixing

[WARNING] Stashed changes conflicted with hook auto-fixes... Rolling back fixes...
An unexpected error has occurred: CalledProcessError: Command: ('/usr/lib/git-core/git', '-c', 'core.autocrlf=false', 'apply', '--whitespace=nowarn', '/home/henniss/.cache/pre-commit/patch1544663784')
Return code: 1
Expected return code: 0
Output: (none)
Errors:
    error: newfile: already exists in working directory


Check the log at /home/henniss/.cache/pre-commit/pre-commit.log
```

`cat newfile` now shows that it is empty. The unstaged changes aren't restored.

**Comment by henniss** [comment]
After posting this, I realized that in fact you don't need any hooks at all. An empty .pre-commit-config.yaml and a single added file will achieve the same result.

**Comment by asottile** [comment]
Looks like an unfortunate bug in git.  These *should* be equivalent but are not:

```console
$ git diff --binary --
diff --git a/t b/t
index e69de29..880d1a7 100644
--- a/t
+++ b/t
@@ -0,0 +1,3 @@
+hi
+hi
+hi
$ git diff-index --binary $(git write-tree) --
diff --git a/t b/t
new file mode 100644
index 0000000..880d1a7
--- /dev/null
+++ b/t
@@ -0,0 +1,3 @@
+hi
+hi
+hi
```

I can probably switch back to `git diff` instead of `git diff-index`. #571 changed this which means the broken behaviour probably regressed in 0.16.0

**Comment by asottile** [comment]
I've made progress by parsing `git status --porcelain -z` and looking at the second character.  need to thread that through though

**Comment by asottile** [comment]
took me a while, but I think I finally figured out a solution -- check out #912

**Comment by asottile** [comment]
This has been released as part of [v1.14.0](https://github.com/pre-commit/pre-commit/releases/tag/v1.14.0) 🎉  -- thanks again!
